Issue-25: Implementation of ReplaceFile action.

Parth-Brahmbhatt opened the pull request.

rdblue commented on core/src/main/java/com/netflix/iceberg/ReplaceFiles.java:
```diff
@@ -16,21 +16,113 @@
+import com.netflix.iceberg.io.OutputFile;
+
+import java.io.IOException;
+import java.util.*;
```
Style: please don't use globbed imports because you can't search for class usage

Parth-Brahmbhatt replied:
Fixed.

rdblue commented on core/src/main/java/com/netflix/iceberg/ReplaceFiles.java:
```diff
@@ -16,21 +16,113 @@
-  protected List<String> apply(TableMetadata base) {
-    return null;
-  }
+    private final TableOperations ops;
```
Style: indentation should be 2 spaces to match the rest of the code.

rdblue commented on core/src/main/java/com/netflix/iceberg/ReplaceFiles.java:
```diff
@@ -16,21 +16,113 @@
+                    try (ManifestWriter writer = new ManifestWriter(manifestReader.spec(), manifestPath, snapshotId())) {
+                        for (ManifestEntry manifestEntry : manifestReader.entries()) {
+                            if (this.pathsToDelete.contains(manifestEntry.file().path()))
+                                if (manifestEntry.status() == ManifestEntry.Status.DELETED) {
```
If you scan use the reader as an `Interable` it will filter out the deleted entries for you. Then you just have to handle the case where some files weren't present in the manifest file.

Parth-Brahmbhatt replied:
I need the manifest entry and there is no helper method for that right now. I am doing the filtering in this class now.

rdblue commented on core/src/main/java/com/netflix/iceberg/ReplaceFiles.java:
```diff
@@ -16,21 +16,113 @@
+                } catch (IOException e) {
+                    throw new RuntimeException("Could not read manifestFile " + manifest);
+                } finally {
+                    addedManifests.add(manifestPath.location());
```
This always adds the new manifest, but if there were no files deleted from the manifest, I think it should reuse the original manifest. This is what happens in [`StreamingDelete`](https://github.com/Netflix/iceberg/blob/master/core/src/main/java/com/netflix/iceberg/StreamingDelete.java#L153) when there are no files deleted. (Eventually, we might avoid writing and deleting the files if the contents can be held in memory.)

Parth-Brahmbhatt replied:
Done.

rdblue commented on core/src/main/java/com/netflix/iceberg/ReplaceFiles.java:
```diff
@@ -16,21 +16,113 @@
+                        }
+                    }
+                } catch (IOException e) {
+                    throw new RuntimeException("Could not read manifestFile " + manifest);
```
There is a `RuntimeIOException` for this purpose.

Parth-Brahmbhatt replied:
Changed to RuntimeIOException.

rdblue commented on core/src/main/java/com/netflix/iceberg/ReplaceFiles.java:
```diff
@@ -16,21 +16,113 @@
+            final List<CharSequence> deletedPaths = Collections.synchronizedList(new ArrayList<>());
+            final List<String> addedManifests = Collections.synchronizedList(new ArrayList<>());
+
+            currentManifests.parallelStream().forEach(manifest -> {
```
Could you use `Tasks` instead of `parallelStream`? That's how other parts of the code parallelize operations and handle errors.

Parth-Brahmbhatt replied:
Done.

rdblue commented on api/src/main/java/com/netflix/iceberg/RewriteFiles.java:
```diff
@@ -30,4 +32,5 @@
  * will throw a {@link CommitFailedException}.
  */
 public interface RewriteFiles extends PendingUpdate<Snapshot> {
+    RewriteFiles replaceFiles(Set<DataFile> filesToDelete, Set<DataFile> filesToAdd);
```
This should have Javadocs.

Also, it seems odd to use both "replace" and "rewrite" in this API. Let's standardize on just "rewrite". `ReplaceFiles` can still be the internal implementation.

Parth-Brahmbhatt replied:
Changed and javadocs added.

rdblue commented on core/src/main/java/com/netflix/iceberg/BaseTable.java:
```diff
@@ -113,7 +113,6 @@ public Rollback rollback() {
     return new RollbackToSnapshot(ops);
   }
 
-
```
Minor: this is a non-functional change.

rdblue commented on core/src/main/java/com/netflix/iceberg/ReplaceFiles.java:
```diff
@@ -16,21 +16,113 @@
+     */
+    @Override
+    protected List<String> apply(TableMetadata base) {
+        ValidationException.check(!(this.pathsToDelete.isEmpty() || this.filesToAdd.isEmpty()), "Must provide some files to add and delete.");
```
`ValidationException` isn't intended for argument validation. Instead, I would suggest using `Preconditions` to throw `IllegalArgumentException`.

`ValidationException` is used for consistency checks, like making sure that a partition spec references valid column ids in a table schema.

Parth-Brahmbhatt replied:
Changed.

rdblue commented on core/src/test/java/com/netflix/iceberg/TestReplaceFiles.java:
```diff
@@ -0,0 +1,244 @@
+  public void testAddOnly() {
+    Assert.assertEquals("Table should start empty", 0, listMetadataFiles("avro").size());
+
+    table.newRewrite()
```
If you use `assertThrows`, then you could also validate the exception's message to make sure it contains a string.

https://github.com/Netflix/iceberg/blob/master/api/src/test/java/com/netflix/iceberg/TestHelpers.java#L112-L131

Parth-Brahmbhatt replied:
Done.

rdblue commented on core/src/test/java/com/netflix/iceberg/TestReplaceFiles.java:
```diff
@@ -0,0 +1,244 @@
+        CommitFailedException.class, "Injected failure", rewrite::commit);
+
+    Assert.assertFalse("Should clean up new manifest", new File(manifest1).exists());
+    Assert.assertFalse("Should clean up new manifest", new File(manifest2).exists());
```
Apply is going to be called each time the commit is reattempted, so it would be good to also check that there aren't any unexpected manifests using `listMetadataFiles`.

Parth-Brahmbhatt replied:
Added.

rdblue commented on core/src/test/java/com/netflix/iceberg/TestReplaceFiles.java:
```diff
@@ -0,0 +1,244 @@
+  }
+
+  @Test
+  public void testAddAndDelete() {
```
This should also verify that there are no unexpected manifests with `listMetadataFiles`.

Parth-Brahmbhatt replied:
Added the check.

Parth-Brahmbhatt commented on core/src/main/java/com/netflix/iceberg/ReplaceFiles.java:
```diff
@@ -16,21 +16,113 @@
+    private final AtomicInteger manifestCount = new AtomicInteger(0);
+    private boolean hasChanges = false;
+
+    public ReplaceFiles(TableOperations ops) {
```
Done

Parth-Brahmbhatt commented on core/src/test/java/com/netflix/iceberg/TestReplaceFiles.java:
```diff
@@ -0,0 +1,244 @@
+
+    rewrite.commit();
+
+    table.newRewrite()
```
I was trying to test 2 cases one where the file existed at one point and got deleted and another case where the file never existed in any manifest. I see the value in the 3rd case that you are describing so will add a test case for that.

Parth-Brahmbhatt commented:
I have addressed most of the comments and I had to refactor the implementation so both RewriteFiles and OverwriteFiles could leverage the same code so it has changed quite a bit. I am going to merge this in to unblock @danielbwatson but feel free to suggest any changes and I can address them in another PR.
